#196: rewrite in TS (closes #196)

FrancescoCioria opened the pull request:
Closes #196

## Test Plan

### tests performed

`yarn test`

giogonzo commented on README.md:
```diff
@@ -4,20 +4,20 @@ A set of shared scripts for your front-end apps.
 ### `metarpheus`
 
-Metarpheus generates tcomb-annotated JS models based on your Scala API. To use,
+Metarpheus generates tcomb-annotated TS models based on your Scala API. To use,
```
io-ts models?

giogonzo commented on README.md:
```diff
@@ -88,30 +79,28 @@ Bundling your application with webpack is awesome. What's less awesome is having
-- a `Config.js` file. It should export a tcomb type validating the configuration. Currently only `port` is strictly required by scriptoni webpack to work
-- any of `production.json`, `development.json`, `local.json` (all are optional): production and development should be tracked in version control, they are the default/base for `NODE_ENV=production` and `=development`, respectively. `local.json` is inteded to be used for custom, per-developer config tweaks, and should not be tracked.
+The `config` directory should include:
+- a `Config.js` file. It should export a `tcomb` type validating the configuration. Currently only `port` is strictly required by scriptoni webpack to work
```
is this still a tcomb type? Or an io-ts one now?

FrancescoCioria replied:
`io-ts`, I'll fix the readme

FrancescoCioria commented on README.md:
````diff
@@ -141,7 +130,7 @@ Let's say, for example, you want to change the output library.
 You can provide a `webpack.config.js` file in the root directory of your project, with the following content:
 
 ```js
-module.exports = (defaultConfig, { config, paths, NODE_ENV, target }) => ({
````
no idea, I didn't remove it in this PR, we should check the history

FrancescoCioria commented on package.json:
```diff
@@ -11,17 +11,15 @@
+    "lint-style-fix": "node bin/scriptoni.js lint-style-fix",
+    "metarpheus": "node bin/scriptoni.js metarpheus",
+    "metarpheus-diff": "node bin/scriptoni.js metarpheus-diff",
+    "test": "NODE_ENV= jest --runInBand",
```
if i'm not wrong, without it `jest` sets `NODE_ENV` to `test`, but we can try removing it as see what happens

FrancescoCioria replied:
yes, it's needed because `scriptoni` defaults `NODE_ENV` to `development` and `production` when running `web-build` and `web-dev` only if it's not defined and as `jest` sets it to `test` those script wouldn't work properly

FrancescoCioria commented on src/scripts/metarpheus/run.ts:
```diff
@@ -0,0 +1,38 @@
+  logger.metarpheus('Starting metarpheus-io-ts');
+  const model = getModels(
+    intermRep.models,
+    { isReadonly: metarpheusConfig.isReadonly, runtime: metarpheusConfig.runtime },
```
I've added [this](https://github.com/buildo/scriptoni/blob/84e5c50a0431afe49692a4c3bac6ee9c31df0213/test/template-app/fake-api-src/Account.scala#L9) `Option` in the fake-api exactly to test that and [it's being serialized with `t.undefined`](https://github.com/buildo/scriptoni/blob/84e5c50a0431afe49692a4c3bac6ee9c31df0213/test/__snapshots__/metarpheus.test.ts.snap#L28-L31) in the snapshot so I think we're fine

giogonzo replied:
yes what I'm saying is that our code treats it as `t.undefined` as you say, but then the scala API at runtime puts `null`s in JSON. Anyway, unrelated to this PR I would say

FrancescoCioria commented on src/scripts/prettier/listDifferent.ts:
```diff
@@ -2,6 +2,6 @@ import execCommand from '../execCommand';
 import getDefaultArgs from './getDefaultArgs';
 
-const defaultArgs = getDefaultArgs('list-different');
+const defaultArgs: any = getDefaultArgs('list-different');
```
mmm you're right, that looks all wrong, the "args" here are the args accepted by any run command, so it should simply be a map with any key...

FrancescoCioria replied:
no it's more complex than that, anyway, I think this util is used to do two different things, that's why the code is hacky... we should refactor that as well

FrancescoCioria commented:
@giogonzo I addressed your comments

The config/options/args logic needs works even after this huge refactor, but there's already a PR for that so could wait for that
